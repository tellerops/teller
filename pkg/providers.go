package pkg

import (
	"fmt"

	"github.com/spectralops/teller/pkg/core"
	"github.com/spectralops/teller/pkg/logging"
	"github.com/spectralops/teller/pkg/providers"
)

type Providers interface {
	GetProvider(name string) (core.Provider, error)
	ProviderHumanToMachine() map[string]string
}

type BuiltinProviders struct {
}

func (p *BuiltinProviders) ProviderHumanToMachine() map[string]string {
	return map[string]string{
		"Heroku":                      "heroku",
		"Vault by Hashicorp":          "hashicorp_vault",
		"AWS SSM (aka paramstore)":    "aws_ssm",
		"AWS Secrets Manager":         "aws_secretsmanager",
		"Google Secret Manager":       "google_secretmanager",
		"Etcd":                        "etcd",
		"Consul":                      "consul",
		".env":                        "dotenv",
		"Vercel":                      "vercel",
		"Azure Key Vault":             "azure_keyvault",
		"Doppler":                     "doppler",
		"CyberArk Conjur":             "cyberark_conjur",
		"Cloudlflare Workers KV":      "cloudflare_workers_kv",
		"Cloudlflare Workers Secrets": "cloudflare_workers_secrets",
		"1Password":                   "1password",
		"Gopass":                      "gopass",
		"LastPass":                    "lastpass",
		"GitHub":                      "github",
		"KeyPass":                     "keypass",
		"FileSystem":                  "filesystem",
		"ProcessEnv":                  "process_env",
	}
}

func (p *BuiltinProviders) GetProvider(name string) (core.Provider, error) { //nolint
	logger := logging.GetRoot().WithField("provider_name", name)
	switch name {
	case "hashicorp_vault":
		return providers.NewHashicorpVault(logger)
	case "aws_ssm":
		return providers.NewAWSSSM(logger)
	case "aws_secretsmanager":
		return providers.NewAWSSecretsManager(logger)
	case "heroku":
		return providers.NewHeroku(logger)
	case "google_secretmanager":
		return providers.NewGoogleSecretManager(logger)
	case "etcd":
		return providers.NewEtcd(logger)
	case "consul":
		return providers.NewConsul(logger)
	case "dotenv":
		return providers.NewDotenv(logger)
	case "vercel":
		return providers.NewVercel(logger)
	case "azure_keyvault":
		return providers.NewAzureKeyVault(logger)
	case "doppler":
		return providers.NewDoppler(logger)
	case "cyberark_conjur":
		return providers.NewConjurClient(logger)
	case "cloudflare_workers_kv":
		return providers.NewCloudflareClient(logger)
	case "cloudflare_workers_secrets":
		return providers.NewCloudflareSecretsClient(logger)
	case "1password":
		return providers.NewOnePassword(logger)
	case "gopass":
		return providers.NewGopass(logger)
	case "lastpass":
		return providers.NewLastPass(logger)
	case "github":
		return providers.NewGitHub(logger)
	case "keypass":
		return providers.NewKeyPass(logger)
	case "filesystem":
		return providers.NewFileSystem(logger)
	case "process_env":
		return providers.NewProcessEnv(logger)
	default:
		return nil, fmt.Errorf("provider '%s' does not exist", name)
	}
}
