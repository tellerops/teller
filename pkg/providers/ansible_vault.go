package providers

import (
	"fmt"
	"os"
	"sort"

	"github.com/joho/godotenv"
	vault "github.com/sosedoff/ansible-vault-go"
	"github.com/spectralops/teller/pkg/core"
	"github.com/spectralops/teller/pkg/logging"
)

type AnsibleVaultClient interface {
	Read(p string) (map[string]string, error)
}

type AnsibleVaultReader struct {
	passPhrase string
}

func (a AnsibleVaultReader) Read(p string) (map[string]string, error) {
	content, err := vault.DecryptFile(p, a.passPhrase)
	if err != nil {
		return nil, err
	}
	return godotenv.Unmarshal(content)
}

type AnsibleVault struct {
	logger logging.Logger
	client AnsibleVaultClient
}

//nolint
func init() {
	metaInto := core.MetaInfo{
		Description:    "Ansible Vault",
		Name:           "ansible_vault",
		Authentication: "ANSIBLE_VAULT_PASSPHRASE.",
		ConfigTemplate: `
  # Configure via environment variables for integration:
  # ANSIBLE_VAULT_PASSPHRASE: Ansible Vault Password

  ansible_vault:
    env_sync:
       path: ansible/vars/vault_{{stage}}.yml

    env:
      KEY1:
        path: ansible/vars/vault_{{stage}}.yml
      NONEXIST_KEY:
        path: ansible/vars/vault_{{stage}}.yml
`,
		Ops: core.OpMatrix{Get: true, GetMapping: true, Put: false, PutMapping: false},
	}
	RegisterProvider(metaInto, NewAnsibleVault)
}

// NewAnsibleVault creates new provider instance
func NewAnsibleVault(logger logging.Logger) (core.Provider, error) {
	ansibleVaultPassphrase := os.Getenv("ANSIBLE_VAULT_PASSPHRASE")
	return &AnsibleVault{
		logger: logger,
		client: &AnsibleVaultReader{
			passPhrase: ansibleVaultPassphrase,
		},
	}, nil
}

// Name return the provider name
func (a *AnsibleVault) Name() string {
	return "AnsibleVault"
}

// Put will create a new single entry
func (a *AnsibleVault) Put(p core.KeyPath, val string) error {
	return fmt.Errorf("provider %q does not implement write yet", a.Name())
}

// PutMapping will create a multiple entries
func (a *AnsibleVault) PutMapping(p core.KeyPath, m map[string]string) error {
	return fmt.Errorf("provider %q does not implement write yet", a.Name())
}

// GetMapping returns a multiple entries
func (a *AnsibleVault) GetMapping(p core.KeyPath) ([]core.EnvEntry, error) {
	// Read existing secret
	a.logger.WithField("path", p.Path).Debug("read secret")
	kvs, err := a.client.Read(p.Path)
	if err != nil {
		return nil, err
	}

	var entries []core.EnvEntry
	for k, v := range kvs {
		entries = append(entries, p.FoundWithKey(k, v))
	}
	sort.Sort(core.EntriesByKey(entries))

	return entries, nil
}

// Get returns a single entry
func (a *AnsibleVault) Get(p core.KeyPath) (*core.EnvEntry, error) { //nolint:dupl
	a.logger.WithField("path", p.Path).Debug("read secret")

	kvs, err := a.client.Read(p.Path)
	if err != nil {
		return nil, err
	}

	k := p.EffectiveKey()
	val, ok := kvs[k]
	if !ok {
		a.logger.WithFields(map[string]interface{}{"path": p.Path, "key": k}).Debug("key not found")
		ent := p.Missing()
		return &ent, nil
	}

	ent := p.Found(val)
	return &ent, nil
}

// Delete will delete entry
func (a *AnsibleVault) Delete(kp core.KeyPath) error {
	return fmt.Errorf("provider %s does not implement delete yet", a.Name())
}

// DeleteMapping will delete the given path recessively
func (a *AnsibleVault) DeleteMapping(kp core.KeyPath) error {
	return fmt.Errorf("provider %s does not implement delete yet", a.Name())
}
