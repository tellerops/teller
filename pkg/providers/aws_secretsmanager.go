package providers

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"sort"

	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/secretsmanager"
	smtypes "github.com/aws/aws-sdk-go-v2/service/secretsmanager/types"
	"github.com/spectralops/teller/pkg/core"
	"github.com/spectralops/teller/pkg/utils"
)

type AWSSecretsManagerClient interface {
	GetSecretValue(ctx context.Context, params *secretsmanager.GetSecretValueInput, optFns ...func(*secretsmanager.Options)) (*secretsmanager.GetSecretValueOutput, error)
	CreateSecret(ctx context.Context, params *secretsmanager.CreateSecretInput, optFns ...func(*secretsmanager.Options)) (*secretsmanager.CreateSecretOutput, error)
	PutSecretValue(ctx context.Context, params *secretsmanager.PutSecretValueInput, optFns ...func(*secretsmanager.Options)) (*secretsmanager.PutSecretValueOutput, error)
}

type AWSSecretsManager struct {
	client AWSSecretsManagerClient
}

func NewAWSSecretsManager() (core.Provider, error) {
	cfg, err := config.LoadDefaultConfig(context.TODO())
	if err != nil {
		return nil, err
	}

	client := secretsmanager.NewFromConfig(cfg)

	return &AWSSecretsManager{client: client}, nil
}

func (a *AWSSecretsManager) Name() string {
	return "aws_secretsmanager"
}

func (a *AWSSecretsManager) GetMapping(p core.KeyPath) ([]core.EnvEntry, error) {
	kvs, err := a.getSecret(p)
	if err != nil {
		return nil, err
	}

	var entries []core.EnvEntry
	for k, v := range kvs {
		entries = append(entries, p.FoundWithKey(k, v))
	}
	sort.Sort(core.EntriesByKey(entries))
	return entries, nil
}

func (a *AWSSecretsManager) Put(kp core.KeyPath, val string) error {
	k := kp.EffectiveKey()
	return a.PutMapping(kp, map[string]string{k: val})
}

func (a *AWSSecretsManager) PutMapping(kp core.KeyPath, m map[string]string) error {
	secrets, err := a.getSecret(kp)
	if err != nil {
		return err
	}

	secretAlreadyExist := len(secrets) != 0

	utils.Merge(m, secrets)
	secretBytes, err := json.Marshal(m)
	if err != nil {
		return err
	}

	secretString := string(secretBytes)
	ctx := context.Background()
	if secretAlreadyExist {
		// secret already exist - put new value
		_, err = a.client.PutSecretValue(ctx, &secretsmanager.PutSecretValueInput{SecretId: &kp.Path, SecretString: &secretString})
		return err
	}

	// create secret
	_, err = a.client.CreateSecret(ctx, &secretsmanager.CreateSecretInput{Name: &kp.Path, SecretString: &secretString})
	if err != nil {
		return err
	}

	return nil
}

func (a *AWSSecretsManager) Get(kp core.KeyPath) (*core.EnvEntry, error) {
	kvs, err := a.getSecret(kp)
	if err != nil {
		return nil, err
	}

	k := kp.EffectiveKey()
	val, ok := kvs[k]
	if !ok {
		ent := kp.Missing()
		return &ent, nil
	}

	ent := kp.Found(val)
	return &ent, nil
}

func (a *AWSSecretsManager) getSecret(kp core.KeyPath) (map[string]string, error) {
	res, err := a.client.GetSecretValue(context.Background(), &secretsmanager.GetSecretValueInput{SecretId: &kp.Path})
	if err != nil {
		var resNotFound *smtypes.ResourceNotFoundException
		if !errors.As(err, &resNotFound) {
			return nil, err
		}

		// doesn't exist - do not treat as an error
		return nil, nil
	}

	if res == nil || res.SecretString == nil {
		return nil, fmt.Errorf("data not found at %q", kp.Path)
	}

	var secret map[string]string
	err = json.Unmarshal([]byte(*res.SecretString), &secret)
	if err != nil {
		return nil, err
	}

	return secret, nil
}
