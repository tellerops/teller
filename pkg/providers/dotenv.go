package providers

import (
	"os"
	"path"
	"sort"

	"github.com/joho/godotenv"
	"github.com/mitchellh/go-homedir"
	"github.com/spectralops/teller/pkg/core"
	"github.com/spectralops/teller/pkg/utils"
)

const (
	filePerm = 0644
	dirPerm  = 0755
)

type DotEnvClient interface {
	Read(p string) (map[string]string, error)
	Write(p string, kvs map[string]string) error
}
type DotEnvReader struct {
}

func (d *DotEnvReader) Read(p string) (map[string]string, error) {
	content, err := os.ReadFile(p)
	if err != nil {
		return nil, err
	}
	return godotenv.Unmarshal(string(content))
}

func (d *DotEnvReader) Write(p string, kvs map[string]string) error {
	content, err := godotenv.Marshal(kvs)
	if err != nil {
		return err
	}
	return os.WriteFile(p, []byte(content), filePerm)
}

type Dotenv struct {
	client DotEnvClient
}

func NewDotenv() (core.Provider, error) {
	return &Dotenv{
		client: &DotEnvReader{},
	}, nil
}

func (a *Dotenv) Name() string {
	return "dotenv"
}

func (a *Dotenv) Put(p core.KeyPath, val string) error {
	k := p.EffectiveKey()
	return a.PutMapping(p, map[string]string{k: val})
}

func (a *Dotenv) PutMapping(kp core.KeyPath, m map[string]string) error {
	p, err := homedir.Expand(kp.Path)
	if err != nil {
		return err
	}

	// check if the file does exist
	_, err = os.Stat(p)
	switch {
	case err == nil:
		// get a fresh copy of a hash
		var into map[string]string
		into, err = a.client.Read(p)
		if err != nil {
			return err
		}
		utils.Merge(m, into)
		return a.client.Write(p, into)
	case os.IsNotExist(err):
		// ensure all subdirectories exist
		err = os.MkdirAll(path.Dir(p), dirPerm)
		if err != nil {
			return err
		}
		return a.client.Write(p, m)
	default:
		return err
	}
}

func (a *Dotenv) GetMapping(p core.KeyPath) ([]core.EnvEntry, error) {
	kvs, err := a.getSecrets(p)
	if err != nil {
		return nil, err
	}
	var entries []core.EnvEntry
	for k, v := range kvs {
		entries = append(entries, p.FoundWithKey(k, v))
	}
	sort.Sort(core.EntriesByKey(entries))
	return entries, nil
}

func (a *Dotenv) Get(p core.KeyPath) (*core.EnvEntry, error) {
	kvs, err := a.getSecrets(p)
	if err != nil {
		return nil, err
	}

	k := p.EffectiveKey()
	val, ok := kvs[k]
	if !ok {
		ent := p.Missing()
		return &ent, nil
	}

	ent := p.Found(val)
	return &ent, nil
}

func (a *Dotenv) getSecrets(kp core.KeyPath) (map[string]string, error) {
	p, err := homedir.Expand(kp.Path)
	if err != nil {
		return nil, err
	}
	return a.client.Read(p)
}
