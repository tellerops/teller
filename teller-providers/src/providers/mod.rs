use std::{collections::HashMap, str::FromStr};

use lazy_static::lazy_static;
use serde::{Deserialize, Serialize};
use serde_variant::to_variant_name;
use strum::{EnumIter, IntoEnumIterator};

#[cfg(test)]
mod test_utils;

#[cfg(feature = "dotenv")]
pub mod dotenv;
pub mod inmem;

#[cfg(feature = "hashicorp_vault")]
pub mod hashicorp_vault;

#[cfg(feature = "ssm")]
pub mod ssm;

#[cfg(feature = "aws_secretsmanager")]
pub mod aws_secretsmanager;

#[cfg(feature = "google_secretmanager")]
pub mod google_secretmanager;

#[cfg(feature = "hashicorp_consul")]
pub mod hashicorp_consul;

#[cfg(feature = "etcd")]
pub mod etcd;

lazy_static! {
    pub static ref PROVIDER_KINDS: String = {
        let providers: Vec<String> = ProviderKind::iter()
            .map(|provider| provider.to_string())
            .collect();
        providers.join(", ")
    };
}
#[derive(
    Serialize, Deserialize, Debug, Clone, Default, PartialOrd, Ord, PartialEq, Eq, EnumIter,
)]
pub enum ProviderKind {
    #[serde(rename = "inmem")]
    Inmem,

    #[default]
    #[cfg(feature = "dotenv")]
    #[serde(rename = "dotenv")]
    Dotenv,

    #[cfg(feature = "hashicorp_vault")]
    #[serde(rename = "hashicorp")]
    Hashicorp,

    #[cfg(feature = "hashicorp_consul")]
    #[serde(rename = "hashicorp_consul")]
    HashiCorpConsul,

    #[cfg(feature = "ssm")]
    #[serde(rename = "ssm")]
    SSM,

    #[cfg(feature = "aws_secretsmanager")]
    #[serde(rename = "aws_secretsmanager")]
    AWSSecretsManager,

    #[cfg(feature = "google_secretmanager")]
    #[serde(rename = "google_secretmanager")]
    GoogleSecretManager,

    #[cfg(feature = "etcd")]
    #[serde(rename = "etcd")]
    Etcd,
}

impl std::fmt::Display for ProviderKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        to_variant_name(self).expect("only enum supported").fmt(f)
    }
}

impl FromStr for ProviderKind {
    type Err = &'static str;

    fn from_str(input: &str) -> Result<Self, Self::Err> {
        let providers = Self::iter()
            .map(|provider| (provider.to_string(), provider))
            .collect::<HashMap<String, Self>>();

        providers.get(input).map_or_else(
            || Err(&PROVIDER_KINDS as &'static str),
            |provider| Ok(provider.clone()),
        )
    }
}
