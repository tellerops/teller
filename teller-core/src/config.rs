use std::cmp::Ordering;
use std::path::PathBuf;
use std::{
    collections::{BTreeMap, HashMap},
    path::Path,
};

use fs_err as fs;
use serde_derive::{Deserialize, Serialize};
use teller_providers::config::{PathMap, ProviderCfg, KV};
use teller_providers::providers::ProviderKind;
use tera::{Context, Tera};

use crate::Result;

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
pub struct Config {
    pub providers: BTreeMap<String, ProviderCfg>,
}

#[derive(Serialize)]
pub struct RenderTemplate {
    pub providers: Vec<ProviderKind>,
}

fn apply_eqeq(config: &mut Config) {
    config.providers.iter_mut().for_each(|(_name, provider)| {
        provider.maps.iter_mut().for_each(|pm| {
            pm.keys.iter_mut().for_each(|(k, v)| {
                // THINK: replace with:
                // 1. templating: {{id}} (identity), {{snake_case}} (snake case it)
                // 2. other symbols: == id, ^^ capitalize, snake case __ lower snake case
                if v == "==" {
                    v.clone_from(k);
                }
            });
        });
    });
}

impl Config {
    /// Config from text
    ///
    /// # Errors
    ///
    /// This function will return an error if serialization fails
    pub fn with_vars(text: &str, vars: &HashMap<String, String>) -> Result<Self> {
        let rendered_text = Tera::one_off(text, &Context::from_serialize(vars)?, false)?;
        let mut config: Self = serde_yaml::from_str(&rendered_text)?;

        apply_eqeq(&mut config);

        Ok(config)
    }

    /// Config from text
    ///
    /// # Errors
    ///
    /// This function will return an error if serialization fails
    pub fn from_text(text: &str) -> Result<Self> {
        Self::with_vars(text, &HashMap::new())
    }

    /// Config from file
    ///
    /// # Errors
    ///
    /// This function will return an error if IO fails
    pub fn from_path(path: &Path) -> Result<Self> {
        Self::from_text(&fs::read_to_string(path)?)
    }

    /// Create configuration template file
    ///
    /// # Errors
    /// When could not convert config to string
    pub fn render_template(data: &RenderTemplate) -> Result<String> {
        let res: BTreeMap<String, ProviderCfg> = data
            .providers
            .iter()
            .map(|p| {
                (
                    format!("{p}_1"),
                    ProviderCfg {
                        kind: p.clone(),
                        maps: vec![PathMap::from_path("example/dev")],
                        ..ProviderCfg::default()
                    },
                )
            })
            .collect();

        let config = Self { providers: res };

        let a: String = serde_yaml::to_string(&config)?;
        Ok(a)
    }
}

#[derive(Debug, Clone, Serialize, Eq, PartialEq)]
pub struct Match {
    pub path: PathBuf,
    pub position: Option<(usize, usize)>,
    pub offset: usize,
    pub query: KV,
}

impl PartialOrd for Match {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Match {
    fn cmp(&self, other: &Self) -> Ordering {
        let query_cmp = self.query.cmp(&other.query);

        if query_cmp != Ordering::Equal {
            return query_cmp;
        }

        self.offset.cmp(&other.offset)
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_yaml_snapshot;

    use super::*;
    #[test]
    fn load_config() {
        std::env::set_var("TEST_LOAD_1", "DEV");
        let config = Config::from_path(Path::new("fixtures/config.yml")).unwrap();
        assert_eq!(config.providers.len(), 2);
        assert_yaml_snapshot!(config);
    }

    #[test]
    fn can_render_template_config() {
        let data = RenderTemplate {
            providers: vec![ProviderKind::Inmem, ProviderKind::Dotenv],
        };

        let config = Config::render_template(&data).unwrap();
        assert_yaml_snapshot!(config);
    }
}
