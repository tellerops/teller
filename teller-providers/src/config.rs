use std::cmp::Ordering;
use std::collections::BTreeMap;

use serde_derive::{Deserialize, Serialize};

use crate::providers::ProviderKind;

fn is_default<T: Default + PartialEq>(t: &T) -> bool {
    t == &T::default()
}

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
pub struct ProviderCfg {
    #[serde(rename = "kind")]
    pub kind: ProviderKind,
    #[serde(rename = "options", skip_serializing_if = "Option::is_none")]
    pub options: Option<serde_json::Value>,
    #[serde(rename = "name", skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    pub maps: Vec<PathMap>,
}

#[derive(Serialize, Deserialize, Debug, Clone, Default, Eq, PartialEq)]
pub enum Sensitivity {
    #[default]
    None,
    Low,
    Medium,
    High,
    Critical,
}

#[derive(Serialize, Deserialize, Debug, Clone, Default, Eq, PartialEq)]
pub struct ProviderInfo {
    pub kind: ProviderKind,
    pub name: String,
}
#[derive(Serialize, Deserialize, Debug, Clone, Default, Eq, PartialEq)]
pub struct PathInfo {
    pub id: String,
    pub path: String,
}

#[derive(Serialize, Deserialize, Debug, Clone, Default, Eq, PartialEq)]
pub struct MetaInfo {
    pub sensitivity: Sensitivity,
    pub redact_with: Option<String>,
    pub source: Option<String>,
    pub sink: Option<String>,
}
#[derive(Serialize, Deserialize, Debug, Clone, Default, Eq, PartialEq)]
pub struct KV {
    pub value: String,
    pub key: String, // mapped-to key
    pub from_key: String,
    pub path: Option<PathInfo>, // always toplevel
    pub provider: Option<ProviderInfo>,
    pub meta: Option<MetaInfo>,
}

impl PartialOrd for KV {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for KV {
    fn cmp(&self, other: &Self) -> Ordering {
        if let (Some(provider), Some(other_provider)) = (&self.provider, &other.provider) {
            let provider_cmp = provider.kind.cmp(&other_provider.kind);
            if provider_cmp != Ordering::Equal {
                return provider_cmp;
            }
        }

        self.key.cmp(&other.key)
    }
}

impl KV {
    #[must_use]
    pub fn to_data(kvs: &[Self]) -> BTreeMap<String, String> {
        let mut data = BTreeMap::new();
        for kv in kvs {
            data.insert(kv.key.clone(), kv.value.clone());
        }
        data
    }

    #[must_use]
    pub fn from_data(
        data: &BTreeMap<String, String>,
        pm: &PathMap,
        provider: &ProviderInfo,
    ) -> Vec<Self> {
        // map all of the data found
        if pm.keys.is_empty() {
            data.iter()
                .map(|(k, v)| Self::from_value(v, k, k, pm, provider.clone()))
                .collect::<Vec<_>>()
        } else {
            // selectively map only keys from pathmap
            pm.keys
                .iter()
                .filter_map(|(from_key, to_key)| {
                    data.get(from_key).map(|found_val| {
                        Self::from_value(found_val, from_key, to_key, pm, provider.clone())
                    })
                })
                .collect::<Vec<_>>()
        }
    }
    #[must_use]
    pub fn from_value(
        found_val: &str,
        from_key: &str,
        to_key: &str,
        pm: &PathMap,
        provider: ProviderInfo,
    ) -> Self {
        Self {
            value: found_val.to_string(),
            key: to_key.to_string(),
            from_key: from_key.to_string(),
            path: Some(PathInfo {
                path: pm.path.clone(),
                id: pm.id.to_string(),
            }),
            provider: Some(provider),
            meta: Some(MetaInfo {
                sensitivity: pm.sensitivity.clone(),
                redact_with: pm.redact_with.clone(),
                source: pm.source.clone(),
                sink: pm.sink.clone(),
            }),
        }
    }
    #[must_use]
    pub fn from_literal(path: &str, key: &str, value: &str, provider: ProviderInfo) -> Self {
        Self {
            value: value.to_string(),
            key: key.to_string(),
            from_key: key.to_string(),
            path: Some(PathInfo {
                id: path.to_string(),
                path: path.to_string(),
            }),
            provider: Some(provider),
            ..Default::default()
        }
    }

    /// represents a KV without any source (e.g. created manually by a user, pending insert to
    /// one of the providers)
    #[must_use]
    pub fn from_kv(key: &str, value: &str) -> Self {
        Self {
            value: value.to_string(),
            key: key.to_string(),
            from_key: key.to_string(),
            ..Default::default()
        }
    }
}

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
pub struct PathMap {
    pub id: String,
    #[serde(rename = "protocol", skip_serializing_if = "Option::is_none")]
    pub protocol: Option<String>,
    #[serde(rename = "path")]
    pub path: String,
    #[serde(default, rename = "keys", skip_serializing_if = "is_default")]
    pub keys: BTreeMap<String, String>,
    #[serde(default, rename = "decrypt", skip_serializing_if = "is_default")]
    pub decrypt: bool,
    #[serde(default, rename = "sensitivity", skip_serializing_if = "is_default")]
    pub sensitivity: Sensitivity,
    #[serde(
        default,
        rename = "redact_with",
        skip_serializing_if = "Option::is_none"
    )]
    pub redact_with: Option<String>,
    #[serde(default, rename = "source", skip_serializing_if = "Option::is_none")]
    pub source: Option<String>,
    #[serde(default, rename = "sink", skip_serializing_if = "Option::is_none")]
    pub sink: Option<String>,
    // ignore population if optional + we got error
    #[serde(default, rename = "optional", skip_serializing_if = "is_default")]
    pub optional: bool,
}

impl PathMap {
    #[must_use]
    pub fn from_path(path: &str) -> Self {
        Self {
            path: path.to_string(),
            ..Default::default()
        }
    }
}
