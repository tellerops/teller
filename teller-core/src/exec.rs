use std::{collections::HashMap, path::Path, process::Output};

// use crate::{Error, Result};
// use teller_providers::errors::{Error, Result};
use crate::{Error, Result};
pub struct Opts<'a> {
    pub pwd: &'a Path,
    pub capture: bool,
    pub sh: bool,
    pub reset_env: bool,
}

const ENV_OK: &[&str] = &[
    "USER",
    "HOME",
    "PATH",
    "TMPDIR",
    "SHELL",
    "SSH_AUTH_SOCK",
    "LANG",
    "LC_ALL",
    "TEMPDIR",
    "TERM",
    "COLORTERM",
    "LOGNAME",
];

/// Run a command
///
/// # Errors
///
/// This function will return an error if running command fails
pub fn cmd(cmdstr: &str, env_kvs: &[(String, String)], opts: &Opts<'_>) -> Result<Output> {
    let words = if opts.sh {
        shell_command_argv(cmdstr.into())
    } else {
        shell_words::split(cmdstr)?.iter().map(Into::into).collect()
    };
    cmd_slice(
        words
            .iter()
            .map(String::as_str)
            .collect::<Vec<_>>()
            .as_slice(),
        env_kvs,
        opts,
    )
}

fn cmd_slice(words: &[&str], env_kvs: &[(String, String)], opts: &Opts<'_>) -> Result<Output> {
    // env handling
    let mut env_map: HashMap<_, _> = if opts.reset_env {
        std::env::vars()
            .filter(|(k, _)| ENV_OK.contains(&k.as_str()))
            .collect()
    } else {
        std::env::vars().collect()
    };

    for (k, v) in env_kvs {
        env_map.insert(k.clone(), v.clone());
    }

    // no shell
    let (first, rest) = words
        .split_first()
        .ok_or_else(|| Error::Message("command has not enough arguments".to_string()))?;

    let mut expr = duct::cmd(Path::new(first), rest)
        .dir(opts.pwd)
        .full_env(&env_map);

    if opts.capture {
        expr = expr.stdout_capture();
    }

    Ok(expr.run()?)
}

#[cfg(unix)]
fn shell_command_argv(command: String) -> Vec<String> {
    use std::env;

    let shell = env::var("SHELL").unwrap_or_else(|_| "/bin/sh".into());
    vec![shell, "-c".into(), command]
}

#[cfg(windows)]
fn shell_command_argv(command: String) -> Vec<String> {
    let comspec = std::env::var_os("COMSPEC")
        .and_then(|s| s.into_string().ok())
        .unwrap_or_else(|| "cmd.exe".into());
    vec![comspec, "/C".into(), command]
}

#[cfg(test)]
mod tests {
    use std::path::Path;

    use insta::assert_debug_snapshot;
    use teller_providers::config::ProviderInfo;
    use teller_providers::config::KV;
    use teller_providers::providers::ProviderKind;

    use super::cmd;
    use super::Opts;

    #[test]
    #[cfg(not(windows))]
    fn run_echo() {
        let out = cmd(
            "echo $MY_VAR",
            &std::iter::once(&KV::from_literal(
                "/foo/bar",
                "MY_VAR",
                "shazam",
                ProviderInfo {
                    kind: ProviderKind::Inmem,
                    name: "test".to_string(),
                },
            ))
            .map(|kv| (kv.key.clone(), kv.value.clone()))
            .collect::<Vec<_>>(),
            &Opts {
                pwd: Path::new("."),
                capture: true,
                reset_env: true,
                sh: true,
            },
        )
        .unwrap();
        let s = String::from_utf8_lossy(&out.stdout[..]);
        assert_debug_snapshot!(s);
    }

    #[ignore]
    #[test]
    fn env_reset() {
        let out = cmd(
            "/usr/bin/env",
            &std::iter::once(&KV::from_literal(
                "/foo/bar",
                "MY_VAR",
                "shazam",
                ProviderInfo {
                    kind: ProviderKind::Inmem,
                    name: "test".to_string(),
                },
            ))
            .map(|kv| (kv.key.clone(), kv.value.clone()))
            .collect::<Vec<_>>(),
            &Opts {
                pwd: Path::new("."),
                capture: true,
                reset_env: false, // <-- notice this!
                sh: false,
            },
        )
        .unwrap();
        let stdout = String::from_utf8_lossy(&out.stdout[..]).to_string();

        // dirty secret here
        assert!(stdout.contains("GITHUB_TOKEN="));

        let out = cmd(
            "/usr/bin/env",
            &std::iter::once(&KV::from_literal(
                "/foo/bar",
                "MY_VAR",
                "shazam",
                ProviderInfo {
                    kind: ProviderKind::Inmem,
                    name: "test".to_string(),
                },
            ))
            .map(|kv| (kv.key.clone(), kv.value.clone()))
            .collect::<Vec<_>>(),
            &Opts {
                pwd: Path::new("."),
                capture: true,
                reset_env: true, // <-- reset env
                sh: false,
            },
        )
        .unwrap();
        let stdout = String::from_utf8_lossy(&out.stdout[..]).to_string();

        assert!(stdout.contains("USER="));
        assert!(stdout.contains("PATH="));
        // no secret here!
        assert!(!stdout.contains("GITHUB_TOKEN="));
    }
}
