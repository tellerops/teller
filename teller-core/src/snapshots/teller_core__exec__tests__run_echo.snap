---
source: src/exec.rs
expression: s
---
"shazam\n"
